crates/storage/provider/src/providers/database.rs

```diff
@@ -555,6 +566,56 @@ where
     }
 }
 
+/// Fetches the Withdrawals that belong to the given block number
+#[inline]
+fn read_transactions_by_number<'a, TX>(
+    tx: &TX,
+    block_number: u64,
+) -> std::result::Result<Option<Vec<TransactionSigned>>, reth_interfaces::db::DatabaseError>
+where
+    TX: DbTx<'a> + Send + Sync,
+{
+    if let Some(body) = tx.get::<tables::BlockBodyIndices>(block_number)? {
+        let tx_range = body.tx_num_range();
+        return if tx_range.is_empty() {
+            Ok(Some(Vec::new()))
+        } else {
+            let mut tx_cursor = tx.cursor_read::<tables::Transactions>()?;
+            let transactions = tx_cursor
+                .walk_range(tx_range)?
+                .map(|result| result.map(|(_, tx)| tx.into()))
+                .collect::<std::result::Result<Vec<_>, _>>()?;
+            Ok(Some(transactions))
+        }
+    }
+
+    Ok(None)
+}
+
+/// Fetches the Withdrawals that belong to the given block number
+#[inline]
+fn read_withdrawals_by_number<'a, TX>(
+    tx: &TX,
+    block_number: u64,
+) -> std::result::Result<Option<Vec<Withdrawal>>, reth_interfaces::db::DatabaseError>
+where
+    TX: DbTx<'a> + Send + Sync,
+{
+    tx.get::<tables::BlockWithdrawals>(block_number).map(|w| w.map(|w| w.withdrawals))
+}
+
+/// Fetches the corresponding header
+#[inline]
+fn read_header<'a, TX>(
+    tx: &TX,
+    block_number: u64,
+) -> std::result::Result<Option<Header>, reth_interfaces::db::DatabaseError>
+where
+    TX: DbTx<'a> + Send + Sync,
+{
+    tx.get::<tables::Headers>(block_number)
```

Comment:
Why are we adding these helpers? Do they provide much? Should probably move to the relevant impl of the HeadersProviders

Reply:
> Should probably move to the relevant impl of the HeadersProviders

this PR does the opposite because having these separetly reduces redundant database transactions by moving the relevant part to read_*(&tx) functions

Reply:
Right but in the couple places they're replaced in this PR they're ready over "tx.get" calls which are already open txs. Which is fine. So in some places the change is not required

Reply:
not required, but otherwise we would have the same chunk of code twice.

Reply:
Alternatively we could just go for fn based helpers instead that take something like `DbTx`/`DbTxMut` (cc @joshieDo) since `Transaction` derefs to that, `DbTx` is obviously already that and `ShareableDatabase` gives you one